### ISSUE TITLE
Background D020 updates late after raster interrupt?

### ISSUE THREAD

**Issue by dansanderson** [MEGA65/mega65-core#803]
A simple assembly language program that sets a raster interrupt for line 200, changes the border (D020) and background (D021) to black (0), counts down from 255 in a short busy-loop, then changes the border and background to white (1) demonstrates that the background does not change until just more than half way across the raster line, for both transitions. The border appears to transition on the first of the two raster lines. This is not explained by CPU cycles spent in the interrupt handler code, at 40 MHz.

I don't actually know if this is unexpected behavior for a C65 or a C64. Is there a known trick to this? Note that at 40 MHz there is no visible "stability" issue often discussed in the context of a C64 1 MHz interrupt.

![Photo on 5-30-24 at 5 38 PM](https://github.com/MEGA65/mega65-core/assets/408973/77c1d69e-d991-4d50-8e5f-e19ad15abbd8)

[raster_hwirq.asm.txt](https://github.com/user-attachments/files/15507042/raster_hwirq.asm.txt)

**Comment by dansanderson** [comment]
I can get the background change position to flicker steadily by eight-ish pixels (just eyeballing) with minor code changes. It would be surprising if this is CPU cycle slop in the interrupt, and seems more likely to be related to when the VIC recognizes a new border color.

I also tried a version where I cycled through 16 colors across 16 adjacent lines, one interrupt per line, and saw an interesting pattern of variability, stable by line position. So that might also suggest the background color change happens at different times based on raster vertical position.

**Comment by ki-bo** [comment]
This is probably related to MEGA65 and no C65 behavior. My feeling is that you see pre-rendered characters where the bg color was still white when the rendering happened (chars are not rendered in real-time, but are rendered into the raster rewrite buffer first, and then those pre-rendered pixels are pushed to the video output later). Rendering of chars probably starts at the end of the previous raster line, and parts of the text are already ready when you change the bg color.

**Comment by ki-bo** [comment]
Can we close this one? I don't see unintended behavior, the VIC behaves as expected.

**Comment by dansanderson** [comment]
The test program attached to this issue (raster_hwirq.asm.txt) produces the same screenshot also attached to the issue above, with core release-0.97,20250422.09,aaf4542.

I think my original concern was that the raster interrupt was not happening when I thought it should be happening. I understand your explanation as, it is happening at the right time, but changing the background color register near this moment doesn't do what I expected it to, because some amount of the line—specifically in the character graphics area—was rendered prior to the interrupt. So the border appears to change on the correct line, but the background appears to change mid-line, with 8 pixels of flicker because that's a pre-rendered character by some amount of timing. All correct?

I wonder how a timing-sensitive graphics program is meant to account for this. But if everyone is comfortable with the way this works, then I'm fine resolving the issue.

**Comment by ki-bo** [comment]
Yes, this is my understanding of how raster rewrite buffer works: It already starts pre-rendering the pixels as soon as possible - which is in the line before they get visible.

> I wonder how a timing-sensitive graphics program is meant to account for this. But if everyone is comfortable with the way this works, then I'm fine resolving the issue.

It indeed makes timing critical code that needs tight raster timing more complicated than on a C64 or C65. The real problem I see with those is rather that you basically can't use significant DMA jobs anymore outside the IRQ. Typically, the jitter when triggering a raster IRQ is defined by waiting for the current CPU instruction to complete, a few cycles. If SEI/CLI protected sections are involved, those can typically be limited, as well. But if you have background jobs copying chunks of memory around with DMA, those will block interrupts potentially for a longer time. Splitting those up into smaller chunks is making things just more complicated again. The C65 has the ability to yield to interrupts during DMA, but the M65 is lacking such a feature, unfortunately. I am not sure whether the traditional 8bit timing-aware programming style of C64 or C65 might be still feasible on the M65 anyway.
